examples/server/server.cpp

```diff
@@ -186,6 +165,212 @@ struct slot_params {
     }
 };
 
+struct server_task {
+    int id    = -1; // to be filled by server_queue
+    int index = -1; // used when there are multiple prompts (batch request)
+
+    server_task_type     type;
+    server_task_inf_type inf_type;
+
+    // used by SERVER_TASK_TYPE_CANCEL
+    int id_target = -1;
+
+    // used by SERVER_TASK_TYPE_INFERENCE
+    slot_params  params;
+    llama_tokens prompt_tokens;
+    int id_selected_slot = -1;
+
+    // used by SERVER_TASK_TYPE_SLOT_SAVE, SERVER_TASK_TYPE_SLOT_RESTORE, SERVER_TASK_TYPE_SLOT_ERASE
+    struct slot_action {
+        int slot_id;
+        std::string filename;
+        std::string filepath;
+    };
+    slot_action slot_action;
+
+    // used by SERVER_TASK_TYPE_METRICS
+    bool metrics_reset_bucket = false;
+
+    server_task(
+        server_task_type type,
+        server_task_inf_type inf_type = SERVER_TASK_INF_TYPE_COMPLETION) : type(type), inf_type(inf_type) {}
```

Comment:
It's better to be explicit here about the inference type:

```suggestion
    server_task(
        server_task_type type,
        server_task_inf_type inf_type) : type(type), inf_type(inf_type) {}
```

Reply:
the inf_type is only required with `SERVER_TASK_TYPE_INFERENCE`, so I think we should provide a default value for it when we don't want to use, right? For example, `server_task(SERVER_TASK_TYPE_METRICS)`

Another way is to remove `server_task_inf_type inf_type` from the constructor and set it later on.

Reply:
The default value should be something like `SEREVR_TASK_INF_TYPE_NONE`.

Though I think it's better to combine the `server_task_type` and `server_task_inf_type` enums in a single enum:

```c++
enum server_task_type {
    SERVER_TASK_TYPE_COMPLETION,
    SERVER_TASK_TYPE_EMBEDDING,
    SERVER_TASK_TYPE_RERANK,
    SERVER_TASK_TYPE_INFILL,
    SERVER_TASK_TYPE_CANCEL,
    SERVER_TASK_TYPE_NEXT_RESPONSE,
    SERVER_TASK_TYPE_METRICS,
    SERVER_TASK_TYPE_SLOT_SAVE,
    SERVER_TASK_TYPE_SLOT_RESTORE,
    SERVER_TASK_TYPE_SLOT_ERASE,
    SERVER_TASK_TYPE_SET_LORA,
};
```

Reply:
I ended up combining them in https://github.com/ggerganov/llama.cpp/pull/10691/commits/090a113417117b90c7c8a946316caa6d9cba7544